core/benches/bench.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
+
+use ink_core::storage2::collections::Stash as StorageStash;
+use num_traits::real::Real;
+
+criterion_group!(benches, criterion_benchmark);
+criterion_group!(benches_worst_case, criterion_benchmark_with_taken_value_read);
+criterion_main!(benches, benches_worst_case);
+
+fn remove_from_filled(test_values: &[u8; 6]) {
+    let mut stash = test_values.iter().copied().collect::<StorageStash<_>>();
+
+    for (index, _value) in test_values.iter().enumerate() {
+        stash.remove(index as u32);
+    }
+    assert_eq!(stash.len(), 0);
+}
+
+fn take_from_filled(test_values: &[u8; 6]) {
+    let mut stash = test_values.iter().copied().collect::<StorageStash<_>>();
+
+    for (index, _value) in test_values.iter().enumerate() {
+        stash.take(index as u32);
+    }
+    assert_eq!(stash.len(), 0);
+}
+
+fn take_from_filled_worst_case(test_values: &[u8; 6]) {
+    let mut stash = test_values.iter().copied().collect::<StorageStash<_>>();
+
+    for (index, _value) in test_values.iter().enumerate() {
+        let val = stash.take(index as u32);
+        let v = val.expect("must exist");
+    }
+    assert_eq!(stash.len(), 0);
+}
```

Comment:
I think there was a misunderstanding here. Let me explain.
In order to enforce that the `storage2::Stash` actually loads the `taken` value from the contract storage we have to instantiate an instance of such a `test_values` storage Stash and then `push_spread` it onto the contract storage using some known `Key`.
When performing the benchmarks we then have to lazily instantiate such a storage Stash using the `pull_spread` method using the same key. It will just load lazily and won't pull anything from the storage, yet. So using `take` from this instance will actually initiate loading from the emulated contract storage. This way we can compare `remove` and `take` operations with the worst-case access for the `take` operation which should yield a significantly higher performance difference since `remove` guarantees to never load from contract storage.

Reply:
Thanks, I added it. Things is because of the issue with the linked list from the comment below the difference currently can't be measured.